### PR TITLE
Feature/lep on retry exceeded

### PR DESCRIPTION
**Pull request by VovaViter**

### CODE REVIEW

File: src/main/java/com/icthh/xm/tmf/ms/activation/service/RetryService.java [hunk 1]
```diff
@@ -1,29 +1,33 @@
...
 
+import javax.annotation.PostConstruct;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.base.Predicates.not;
+import static com.icthh.xm.tmf.ms.activation.domain.SagaEvent.SagaEventStatus.*;
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/tmf/ms/activation/service/RetryService.java:26]: replace wildcard import to direct imports

File: src/test/java/com/icthh/xm/tmf/ms/activation/config/TenantConfigMockConfiguration.java [hunk 2]
```diff
@@ -67,9 +63,9 @@ public TenantVerifyInterceptor tenantVerifyInterceptor() {
         return mock(TenantVerifyInterceptor.class);
     }
 
-    @Bean
+   /* @Bean
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/tmf/ms/activation/config/TenantConfigMockConfiguration.java:66]: remove commented code

File: src/test/resources/lep/RetryLimitExceeded$$around.groovy [hunk 3]
```diff
@@ -0,0 +1,3 @@
+def event = lepContext.inArgs.sagaEvent;
+event.taskContext.put('test', 'data');
+return [:]
```
Comment (sergeysenja1992) [inline src/test/resources/lep/RetryLimitExceeded$$around.groovy:3]: add new line

File: src/test/resources/spec/activation-spec-retry-service-test.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+---
+transactions:
+  - key: TEST-TYPE-KEY
+    retryPolicy: RETRY
+    retryCount: 3
+    tasks:
+      - key: TASK-1
+        retryCount: 3
+        maxBackOff: 0
```
Comment (sergeysenja1992) [inline src/test/resources/spec/activation-spec-retry-service-test.yml:9]: add new line

File: src/test/java/com/icthh/xm/tmf/ms/activation/service/RetryServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            SagaEvent event = (SagaEvent) invocation.getArguments()[0];
+            retryService.retry(event, task, ON_RETRY);
+            countDownLatch.countDown();
+            return null;
+        }).when(eventsSender).sendEvent(refEq(sagaEvent, "backOff", "retryNumber", "status"));
+
+        retryService.retry(sagaEvent, task, ON_RETRY);
+        countDownLatch.await(5, TimeUnit.SECONDS);
+
+        verify(eventRepository, times(6)).save(refEq(sagaEvent, "backOff", "retryNumber", "status"));
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/tmf/ms/activation/service/RetryServiceTest.java:146]: replace to verify save without ignore retryNumber and status

File: src/test/java/com/icthh/xm/tmf/ms/activation/service/RetryServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        verify(eventRepository, times(3)).findById(eq(id));
+
+        Assert.assertThat(sagaEvent.getTaskContext(), IsMapContaining.hasEntry("test", "data"));
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/tmf/ms/activation/service/RetryServiceTest.java:149]: add noMoreInteraction assert for each mock-s

File: src/main/java/com/icthh/xm/tmf/ms/activation/service/RetryService.java [hunk 7]
```diff
@@ -54,8 +58,13 @@ public void retry(SagaEvent sagaEvent, SagaTaskSpec sagaTaskSpec, SagaEvent.Saga
...
 
         if (sagaEvent.getRetryNumber() > sagaTaskSpec.getRetryCount() && sagaTaskSpec.getRetryCount() >= 0) {
-            log.warn("Limit retry exceeded for event {}. {} > {}", sagaEvent, sagaEvent.getRetryNumber(),
-                     sagaTaskSpec.getRetryCount());
+            log.warn("Retry limit exceeded for event {}. {} > {}", sagaEvent, sagaEvent.getRetryNumber(),
+                sagaTaskSpec.getRetryCount());
+            try {
+                self.retryLimitExceeded(sagaEvent, sagaTaskSpec, eventStatus);
+            } catch (Throwable e) { // Because of fact that groovy code can have compilation errors
+                log.error(" Error unable to start compensation lep ", e);
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/activation/service/RetryService.java:66]: suggest changing log to :
`log.error("Error unable to start compensation lep: {}", e.getMessage(), e);`

File: src/test/java/com/icthh/xm/tmf/ms/activation/service/RetryServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        });
+
+        sagaSpecService.onRefresh("/config/tenants/XM/activation/activation-spec.yml", loadFile("spec/activation-spec-retry-service-test.yml"));
+
+    }
+
+    @SneakyThrows
+    @Test
+    public void testRetry() {
+        lepResourceLoader.onRefresh("/config/tenants/XM/activation/lep/service/retry/RetryLimitExceeded$$around.groovy", loadFile("/lep/RetryLimitExceeded$$around.groovy"));
```
Comment (amedvedchuk) [inline src/test/java/com/icthh/xm/tmf/ms/activation/service/RetryServiceTest.java:114]: should we add test for lep attached by Task type key as well?

**Pull request merged**